### ISSUE TITLE
PUT behaviour

### ISSUE THREAD

**Issue by cportele** [opengeospatial/ogcapi-styles#40]
There are two options how to handle PUT on `{baseResource}/styles/{styleId}`, depending on whether the server supports automated derivation of stylesheets in other style encodings.

1. The PUT request removes all existing stylesheets of the style. The submitted stylesheet is accepted as the new style definition. If the server supports automatic derivation of stylesheets in other style encodings, it will derive other stylesheets from the new style definition and add them to the style metadata.

2. The PUT request sets or replaces the current stylesheet for the media type specified in the Content-Type header, but leaves other stylesheets of the style untouched. That is, all stylesheets are managed separately by clients.

Do we want to support one or both of the options? How do clients know the behaviour of the API?

**Comment by cportele** [comment]
Meeting 2021-11-02: 

* This is only an issue for clients that manage styles and stylesheets, not for clients that simply fetch styles.
* It is not an issue specific to styles, it can also affect other resource.
* Conditional headers could help here to avoid unintentional overrides. Similar to the case with Features where two clients may update the same feature and the conditional headers can avoid this. @cportele will draft a description how this capability can be used for these cases, too.

**Comment by jerstlouis** [comment]
@cportele Which headers did you have in mind that could help with this? In Part 4, I see mentions of If-Match and If-Unmodified-Since, was it one of those?

**Comment by cportele** [comment]
@jerstlouis 
Yes, that was the point. That aspect of the issue is now obsolete (as far as Styles is concerned), since we have added the relevant requirements classes in Features Part 4 (in 2022).

**Comment by jerstlouis** [comment]
Thanks @cportele . Not sure I fully understand how these headers help this particular use case...

Is it that a client first uploads a Mapbox GL style, and then it also uploads an SLD/SE style but it uses If-Unmodified-Since e.g., the time that it sent the SLD/SE, so that if the server automatically derived one from the Mapbox GL one it previously uploaded it cancels that operation?

If the server does not support storing multiple independent formats for the same style (it has a single internal representation), then I guess a client wishing to itself upload multiple representations simply cannot do that, and other than the documentation the server cannot really advertise whether it supports this or not? But at least it can first upload its "best" representation first, and then use the conditional header for the alternate ones. It would still be useful to include this guidance I think.

Regarding this being obsolete, I guess related to this we need to consider referencing Features - Part 4 / CRUD (did we say this could also be dual-title Common part?).

**Comment by cportele** [comment]
@jerstlouis 
I don't think this is directly related to the main issue. I guess at the time we had a general discussion of PUT semantics and this topic came up, too.

Regarding the original issue: In my view, if the style format is not part of the URI, then only option 1 is consistent with the HTTP semantics as the URI identifies a resource. If I want to manage all representations / style formats of the style separately, each format needs its own URI. Regarding CRUD operations, the API will declare which formats it accepts. 

I am not really sure, if there is a need support for a capability as described in option 2. I think that is more likely an edge case and could also be handled in the future, if there is really a need.

Features Part 4 is already [a normative reference](https://docs.ogc.org/DRAFTS/20-009.html#_references). What needs to be added are conditional dependencies to the optimistic locking requirements classes. And yes, the idea is to co-brand that standard as well as all those requirements classes are not at all specific to features as resources.
